### ISSUE TITLE
Feature Request: Tooltips for Timeline Tags?

### ISSUE THREAD

**Issue by AVLinux** [hydrogen-music/hydrogen#1807]
Hydrogen version *          : 1..2.0
Operating system + version  : Linux  AV Linux 21.3 (Debian-based)
Audio driver + version      :  PulseAudio

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Hi, after quite a few years away from Hydrogen I am back with a vengeance using it for live performances with an Acoustic Duo and it is working very well with a hand-percussion based kit I'm still working on but will share with the community when it is complete... Anyway... I know the 'Director' exists and it is pretty handy but I personally make extensive use of the Timeline tagging feature and for me when following along with a song arrangement it would be pretty cool if there were Tooltips of the Tags that popped up when the playhead passes the Tags, this would be an easy way to see each part in the timeline View at a glance without using the large Director window.

A second note about that... I understand the need for the Director Window to be seen clearly but on a small laptop screen it obscures the timeline View and I find seeing the blocks on the timeline is also a useful way to get positional information so it might be nice to have a 'Director Large' and 'Director Small' Window option so people can see both if wanted..

Lastly while I'm thinking of it, when performing live it is a pain to have to hit Alt-D in a huge rush when you have started your song to make the Director Window appear, is there a way that it can just stay visible when you click to start the song? 

Thanks for the continuing work on Hydrogen, I don't know how many people perform live with it but it is a great use for an old laptop and infinitely more flexible than pedals like the 'Beat Buddy' and the various loopers with internal Drum beats etc..

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

*  H2 version : see 'Info - About', or in case you build Hydrogen from source use 'git rev-parse HEAD'
** if you are reporting a bug, please describe how the bug can be reproduced

**Comment by theGreatWhiteShark** [comment]
> I know the 'Director' exists and it is pretty handy but I personally make extensive use of the Timeline tagging feature and for me when following along with a song arrangement it would be pretty cool if there were Tooltips of the Tags that popped up when the playhead passes the Tags, this would be an easy way to see each part in the timeline View at a glance without using the large Director window.
>
*added note: It would also be handy if the Tooltips would appear a bar or a certain number of beats before the Playhead actually passed the Tag that way you would get a heads-up for what part is coming next like the Director does.

I like that idea a lot! Personally, I never used Tags because they are pretty useless outside of the Director. That's why I gave them some work in version 1.2. to at least allow them to coexist neatly with our Tempo Markers and show their content in a tooltip while hovering.

The Timeline itself is already packed with loads of information and the column numbers below are too important to be covered by a tooltip. Maybe the upper part of the Timeline right above the Tempo Marker and some additional space in the lower part of the PlayerControl (with the play, stop etc. buttons) would work? I'll check.

I will also add an option to the Preferences to turn off these tooltips for users preferring the "old" UX.

I'm still a bit puzzled about the way to present the tooltips. I tend towards showing the last Tag of the current transport position (the one reached first when moving from the playhead to the left) statically. So, whenever playhead is located between Tag A and B - even if transport is not rolling - A is displayed in a tooltip right above the column of A.

The upcoming Tag could be shown whenever the playhead enters the column left of it (B - 1). But it is still shown at column B. For long/close Tags it would cover parts of A without any transparency and maybe have a slightly different color in order to indicate that it is a preview of an upcoming Tag.

What do you think? Does this UX makes sense?

> A second note about that... I understand the need for the Director Window to be seen clearly but on a small laptop screen it obscures the timeline View and I find seeing the blocks on the timeline is also a useful way to get positional information so it might be nice to have a 'Director Large' and 'Director Small' Window option so people can see both if wanted..

Ideally, at some distant point in time, the individual parts of the GUI could be docked and undocked to e.g. have a smaller director or playlist editor as well as to have a larger Sound Library or Automation Path. But there are still loads of more pressing things to do and it's a rather larger task.

> Lastly while I'm thinking of it, when performing live it is a pain to have to hit Alt-D in a huge rush when you have started your song to make the Director Window appear, is there a way that it can just stay visible when you click to start the song?

For most dialogs we record and store the windows visibility, geometry, and position in the `~/.hydrogen/hydrogen.conf` file. But you are right, not for the Director. I'll add this.

In addition, the future 1.3.0 version will allow you customize all keyboard bindings and also to create new ones triggering multiple actions. You could e.g. bind "Show Director" and "Toggle Playback" to <kbd>Alt</kbd> + <kbd>Space</kbd>. This should already work on `master`.

**Comment by AVLinux** [comment]
Hi,

Thanks for the quick reply and even humoring the idea of implementing this, it's very much appreciated!

I think your UX ideas are very good and using columns rather than beats or bars to trigger the Tooltip part name changes would be perfect! Also the Tooltip keeping the name of the last tag passed is a good idea also.

I understand your concern about where the Tooltips themself will appear but I would suggest that perhaps now that you have Tags and tempo info sitting on top of each other maybe the timeline would benefit from becoming a few pixels vertically higher in general since it is such a dominant functional part of the UI? I know folks with small screens never want UI items to get bigger but the Timeline is where so much positional and active information is happening, is there room for a small increase?

**Comment by theGreatWhiteShark** [comment]
> I understand your concern about where the Tooltips themself will appear but I would suggest that perhaps now that you have Tags and tempo info sitting on top of each other maybe the timeline would benefit from becoming a few pixels vertically higher in general since it is such a dominant functional part of the UI? I know folks with small screens never want UI items to get bigger but the Timeline is where so much positional and active information is happening, is there room for a small increase?

Yeah, you are right. For tags to be really handy and useful the Timeline should provide first class tag support via a dedicated row. Then we do not have to deal with tooltips and where they should appear but use a bigger version of the current "T" boxes as a placeholder which expand to boxes containing the actual tag when reached by the playhead.

But that will be some work. I hope to make this part of v1.3 but there are a couple of other things that need to be done first.

> As far as the Director changes the things you already have in the works are probably sufficient, and if the conf file is now going to remember the Director size and position that is a good solution..

Could you check #1810? This should make Hydrogen remember visibility of the Director.

>  Thinking about it further if the tooltip thing comes to fruition the importance of the Director will diminish at least for those who embrace the tooltip functionality..

Hmm... A full-screen Director may still prove useful during live performances in case you place your laptop a couple of meters away.
